### DISCUSSION TITLE
Allow Parameterizing Routes in routes.php

### DISCUSSION THREAD

**Discussion by engram-design** [craftcms/cms#10789] [Ideas]
We often want to setup routes for account pages, but because these aren't entries, we need to setup custom routes. Take the following route rules into consideration for being able to edit a Campaign on the front-end.

```
return [
    'campaigns/edit/<slug:(?:[^\/]*)?>' => ['template' => 'campaigns/edit'],
    'campaigns/edit/<slug:(?:[^\/]*)?>/profile' => ['template' => 'campaigns/edit/profile'],
    'campaigns/edit/<slug:(?:[^\/]*)?>/subscriptions' => ['template' => 'campaigns/edit/subscriptions'],
    'campaigns/edit/<slug:(?:[^\/]*)?>/donate' => ['template' => 'campaigns/edit/donate'],
    'campaigns/edit/<slug:(?:[^\/]*)?>/notifications' => ['template' => 'campaigns/edit/notifications'],
];
```

Here, we want to allow user to edit their campaign from the front end, hence the `campaigns/edit` template. But because a campaign has **lots** of content, we want to break it up into different templates and routes. You can see these additional templates.

What would be amazing is to support [Yii's Parameterizing Routes](https://www.yiiframework.com/doc/guide/2.0/en/runtime-routing#parameterizing-routes) with templates, as they do for controller actions.

Then we could do something far more efficient like:

```
return [
    'campaigns/edit/<slug:(?:[^\/]*)?>' => ['template' => 'campaigns/edit'],
    'campaigns/edit/<slug:(?:[^\/]*)?>/<page:(?:[^\/]*)?>' => ['template' => 'campaigns/edit/<page>'],
];
```

Is this possible now, and if not - do you think this makes valid sense? Or just let me know if I'm going about this the wrong way.

**Comment by brandonkelly** [comment]
Yes that’s already possible. Any params that were matched from the URL pattern will be available as template variables, e.g. `slug`.

**Reply by engram-design** [reply]
Right, so we'd need to use rules like:

```
return [
    'campaigns/edit/<slug:(?:[^\/]*)?>' => ['template' => 'campaigns/edit'],
    'campaigns/edit/<slug:(?:[^\/]*)?>/<page:(?:[^\/]*)?>' => ['template' => 'campaigns/_page'],
];
```

With the content of `campaigns/_page` being:

```
{% include ['campaigns/edit/' ~ page] %}
```

We can't have it route to the correct template based on the route params without the intermediary Twig template?

**Reply by brandonkelly** [reply]
Ohhh sorry I wasn’t really following the question, and missed the `<page>` bit in the example. Not currently possible. Will unmark this as the answer and relabel this as an idea.

**Reply by engram-design** [reply]
All good! I know it's pretty niche, but something we've run across on a few projects that would help clean things up (and according to the Yii docs, help the URL Manager performance). But my approach above is probably the way to go

**Comment by mmikkel** [comment]
I'd really love to see this implemented!
